java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/project/LspBrokenReferencesImpl.java

```diff
@@ -0,0 +1,161 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.java.lsp.server.project;
+
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.api.project.Project;
+import org.netbeans.spi.project.ui.ProjectProblemsImplementation;
+
+/**
+ * Implements the ProjectProblem resolution process in a way more suitable for LSP. When a project Alert
+ * or Customizer is requested, a ProjectAlertPresenter is created and will be live until the user processes
+ * all the alerts, or the responses time out. With a new alert, the alerts will show up again.
+ * Each Presenter works with a single Project.
+ * <p>
+ * When fatal (non-resolvable) problems occur, they are reported using ShowMessage requests. Once they are
+ * confirmed, resolvable issues will be processed. There's a timeout {@link #WAKEUP_DELAY} for the user
+ * inactivity after last user's response (message ok/dismiss): if no response is received, the process
+ * continues.
+ * <p>
+ * If there are recoverable problems, the user is presented with the first one's description. If there are
+ * more problems, a note "There are XX other fixable problems." is added.
+ * <p>
+ * The main loop supports two modes: autoresolve true will immediately attempt to resolve 1st resolvable
+ * problem, while autoresolve false provides the description and asks for confirmation. This allows the
+ * "OK" and "Rest" button to switch between the two modes:
+ * <ul>
+ * <li>resolve first, report issues if any
+ * <li>ask with details about the fix first, then resolve
+ * </ul>
+ * When the user is asked to decide, a timeout task is scheduled: if the user does not respond within
+ * the defined timeout, the process continues.
+ * <p>
+ * To avoid possible issues with multiple project actions, <b>all resolve calls from all ProblemResolvers</b>
+ * are serialized to a single dedicated RP.
+ * <p/>
+ * A request to <b>alert</b> the project will just push the process further: if timeout is under way, it will 
+ * terminate it and display the next question.
+ * <p/>
+ * A request to <b>display a customizer</b> will invalidate the current presenter, and display all the issues
+ * anew. This allows for an explicit CodeAction or Command that will reiterate questions the user may have closed.
+ * <p/>
+ * Note: if a ShowMessage request is delivered to vscode client and the same ShowMessage request is pending, the client 
+ * will cancel the former ShowMessage as if the user pressed ESC or dismissed the message. The new ShowMessage will be 
+ * displayed
+ * @author sdedic
+ */
+// BAD BAD BAD - DO NOT REGISTER THIS IN LSP SERVER, put it in the integration module !!!
```

Comment:
Would making this class `abstract` with `protected` constructor help to remove the `BAD BAD` message?

Reply:
But how can the integration module access this class, when it is not in public package?

Reply:
It would not :) it's there just for case that someone won't get the great idea to add `@ServiceProvider` since the lsp.server module is a part of NetBeans IDE distro and this override would screw the IDE up.

The integration module uses an explicit META-INF/Services registration that references this class by name. For that the default constructor has to be public. If too ugly, the alternative would be to expose a factory method for the implementation as an API, but I really do not want to expose the implementation class in the APIs.

Reply:
There is a common pattern used for this kind of registrations and there are multiple [AbstractLspXYZ](https://github.com/apache/netbeans/tree/8789e44fef1d416288f6fb62ec9a79442dee9f99/java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/ui) classes. What is so special on this class that requires a special treatment?

Unless you have some intrinsic answer to the above question, then I **suggest** to create `AbstractLspProjectProblemsImplementation`, make it `abstract`, have `protected` constructor like other implementations do. However, it is just a **suggestion**, do as you like.

Reply:
Dear colleague(s). Your peristence in commenting my laziness finally worked... So this should be fixed in 39596e2d6f